Libraries/LibGfx/JPGLoader.cpp

```diff
@@ -0,0 +1,1057 @@
+                auto& previous_dc = context.previous_dc_values[cindex];
+                select_component[0] = previous_dc += dc_diff;
+
+                // Compute the AC coefficients.
+                for (int j = 1; j < 64;) {
+                    symbol_or_error = get_next_symbol(context.huffman_stream, ac_table);
+                    if (!symbol_or_error.has_value())
+                        return false;
+
+                    // AC symbols encode 2 pieces of information, the high 4 bits represent
+                    // number of zeroes to be stuffed before reading the coefficient. Low 4
+                    // bits represent the magnitude of the coefficient.
+                    auto ac_symbol = symbol_or_error.release_value();
+                    if (ac_symbol == 0) {
+                        for (; j < 64; j++)
+                            select_component[zigzag_map[j]] = 0;
+                        continue;
+                    }
+
+                    // ac_symbol = 0xF0 means we need to skip 16 zeroes.
+                    u8 run_length = ac_symbol == 0xF0 ? 16 : ac_symbol >> 4;
+
+                    if (j + run_length >= 64) {
+                        dbg() << String::format("Run-length exceeded boundaries. Cursor: %i, Skipping: %i!", j, run_length);
+                        return false;
+                    }
+
+                    for (int k = 0; k < run_length; k++)
+                        select_component[zigzag_map[j++]] = 0;
+
+                    u8 coeff_length = ac_symbol & 0x0F;
+                    if (coeff_length > 10) {
+                        dbg() << String::format("AC coefficient too long: %i!", coeff_length);
+                        return false;
+                    }
+
+                    // If coeff_length == 0, we need to read another symbol to decode the coefficient.
+                    if (coeff_length != 0) {
+                        coeff_or_error = read_huffman_bits(context.huffman_stream, coeff_length);
+                        if (!coeff_or_error.has_value())
+                            return false;
+                        i32 ac_coefficient = coeff_or_error.release_value();
+                        if (ac_coefficient < (1 << (coeff_length - 1)))
+                            ac_coefficient -= (1 << coeff_length) - 1;
+
+                        select_component[zigzag_map[j++]] = ac_coefficient;
+                    }
+                }
+            }
+        }
+    }
+
+    return true;
+}
+
+Optional<Vector<Macroblock>> decode_huffman_stream(JPGLoadingContext& context)
+{
+    Vector<Macroblock> macroblocks;
+    macroblocks.resize(context.mblock_meta.padded_total);
+
+    dbg() << "Image width: " << context.frame.width;
+    dbg() << "Image height: " << context.frame.height;
+    dbg() << "Macroblocks in a row: " << context.mblock_meta.hpadded_count;
+    dbg() << "Macroblocks in a column: " << context.mblock_meta.vpadded_count;
```

Comment:
This will be more useful when everything's ready for production, but placing your "normal behaviour" debugging in a `#ifdef JPG_DEBUG / #endif` block already helps separating it from error-path debugging.
You can just keep a `#define JPG_DEBUG` command at the top of the file while you need to see the output, it just makes it easier to disable down the line ^^